src/main/java/org/checkerframework/languageserver/CFTextDocumentService.java

```diff
@@ -152,15 +172,100 @@ public void didSave(DidSaveTextDocumentParams params) {
                 Collections.singletonList(new File(URI.create(params.getTextDocument().getUri()))));
     }
 
+    /** Diagnostic Kind NOTE is used for Type Refinements */
     @Override
     public void publish(Map<String, List<javax.tools.Diagnostic<?>>> result) {
         for (Map.Entry<String, List<javax.tools.Diagnostic<?>>> entry : result.entrySet()) {
-            server.publishDiagnostics(
-                    new PublishDiagnosticsParams(
-                            entry.getKey(),
-                            entry.getValue().stream()
-                                    .map(this::convertToLSPDiagnostic)
-                                    .collect(Collectors.toList())));
+            String currentKey = convertEntryKeyToDocURI(entry.getKey());
+            List<Diagnostic> diagnostics = new ArrayList<>();
+            entry.getValue().stream()
+                    .forEach(
+                            i -> {
+                                if (i.getKind() == javax.tools.Diagnostic.Kind.NOTE) {
+                                    publishTypeRefinementhHelper(i, currentKey);
```

Comment:
I think one of the comments was, although a type refinement message is always a `NOTE`, the reversed statement is not always true. So you should probably check if the message contains "[type.refinements]", and you have two options:
1. check before calling `publishTypeRefinementhHelper`.
2. change `publishTypeRefinementhHelper` to `processNotes` and only process messages with "[type.refinements]" for now.